<a href="https://colab.research.google.com/github/bimbeo/Web/blob/master/Excises1_1_35.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**1, In một chuỗi trong một định dạng cụ thể**

**Write a Python program to print the following string in a specific format (see the output).**



Bài đơn giản, chỉ đơn giản là in 1 chuỗi. Sử dụng /n để xuống dòng, /t để cách 1 dấu tab![alt text](https://www.w3resource.com/w3r_images/python-basic-image-exercise-1.png)

In [0]:
print("Twinkle, twinkle, little star, \n\tHow I wonder what you are! \n\t\tUp above the world so high, \n\t\tLike a diamond in the sky. \nTwinkle, twinkle, little star, \n\tHow I wonder what you are!")


Twinkle, twinkle, little star, 
	How I wonder what you are! 
		Up above the world so high, 
		Like a diamond in the sky. 
Twinkle, twinkle, little star, 
	How I wonder what you are!


**2, Nhận phiên bản Python**

**Write a Python program to get the Python version you are using.**



Mô-đun 'sys' cung cấp quyền truy cập vào một số biến được trình thông dịch sử dụng hoặc duy trì và các chức năng tương tác mạnh với trình thông dịch

In [0]:
import sys
print("Python version", sys.version)
print("Version info.", sys.version_info)


Python version 3.6.9 (default, Apr 18 2020, 01:56:04) 
[GCC 8.4.0]
Version info. sys.version_info(major=3, minor=6, micro=9, releaselevel='final', serial=0)


**3,Hiển thị ngày giờ hiện tại .**

**Write a Python program to display the current date and time.**


Hiển thị thời gian hiện: import module date time, thao tác ngày, thời gian

datetime.now (tz = Không) trả về ngày và giờ địa phương hiện tại. Nếu đối số tùy chọn tz là Không hoặc không được chỉ định, thì điều này giống như ngày hôm nay ().

date.strftime (format) trả về một chuỗi đại diện cho ngày, được điều khiển bởi một chuỗi định dạng rõ ràng. Mã định dạng đề cập đến giờ, phút hoặc giây sẽ thấy 0 giá trị. 

In [0]:
import datetime
now = datetime.datetime.now()
print( ("Current date and time : "))
print (now.strftime("%A-%Y-%m-%d %H:%M:%S-%Z")) # Year, Months, day, hour, minute, seconds
print(now)
print(str(now))
print(repr(now)) # trả về đầy đủ hơn

Current date and time : 
Wednesday-2020-05-13 12:37:28-
2020-05-13 12:37:28.675431
2020-05-13 12:37:28.675431
datetime.datetime(2020, 5, 13, 12, 37, 28, 675431)


 **4, Tính diện tích hình tròn: import pi từ module math. Viết hàm tính diện tích pi*r bình phương**

**Write a Python program to display the current date and time.**


 ![alt text](https://www.w3resource.com/w3r_images/python-area-of-a-circle-exercise-4.png)


In [0]:
from math import pi
r = float(input("Input the radius of the circle: "))
print("The area of the circle with radius "+ str(r) + " is " + str(pi*r**2))

Input the radius of the circle: 5
The area of the circle with radius 5.0 is 78.53981633974483


**5, In tên họ, tên người dùng được nhập từ bàn phím và đảo ngược thứ tự, cách nhau bởi khoảng cách ở giữa**

**Write a Python program which accepts the user's first and last name and print them in reverse order with a space between them.**
![alt text](https://www.w3resource.com/w3r_images/python-basic-image-exercise-5.png)

In [0]:
fname = input("Input your First Name: ")
lname = input("Input your Last Name: ")
print("Hello " + lname + " " + fname)

Input your First Name: Bim
Input your Last Name: Beo
Hello Beo Bim


**6, Tạo 1 list và tuple từ những số cách nhau bởi dấu phẩy**

**Write a Python program which accepts a sequence of comma-separated numbers from user and generate a list and a tuple with those numbers**.



In [0]:
values = input("Input some comma seprated numbers : ") # lấy 1 dãy số nhập từ người dùng
list = values.split(",") # tạo 1 list, gán số từ người dùng vào list
tuple = tuple(list) # tạo 1 tuple, gán nội dung từ list vào tuple
print('List : ',list) # in
print('Tuple : ',tuple) # chạy trên colab bị lỗi

Input some comma seprated numbers : 1


TypeError: ignored

**7, Tạo 1 file name được nhập từ người dùng và in ra phần mở rộng của nó. Ví dụ Bimbeo.doc thì in ra doc**

**Write a Python program to accept a filename from the user and print the extension of that.**

![alt text](https://www.w3resource.com/w3r_images/python-basic-image-exercise-7.png)

In [0]:
filename = input("Input the Filename: ")
f_extns = filename.split('.') # tạo 1 list mới, cắt chuỗi được nhập vào được ngăn cách bằng dấu chấm, bim.exe thì cắt thành bim và exe
print("The extension of the file is : " + repr(f_extns[-1])) # in ra index -1 của chuỗi, sử dụng repr để chuyển dữ liệu thành str, nhưng có thêm dấu nháy

Input the Filename: bim.111
The extension of the file is : '111'


**8, Hiển thị đầu và cuối của color từ list colors đã cho**

**Write a Python program to display the first and last colors from the following list.**


![alt text](https://www.w3resource.com/w3r_images/python-basic-image-exercise-8.png)

In [0]:
color_list = ["Red", "Green", "White", "Black"]
print(color_list[0], color_list[-1]) # in index bình thường
print("%s %s"%(color_list[0], color_list[-1])) # in index nhưng sử dụng phép gán %s với định dạng format

Red Black
Red Black


**9, Viết 1 chương trình để hiển thị lịch thi**

**Write a Python program to display the examination schedule. (extract the date from exam_st_date).**


![alt text](https://www.w3resource.com/w3r_images/python-basic-image-exercise-9.png)

In [0]:
exam_st_date = (13,5,2020)
print("The examination will start from: %i/%i/%i"%(exam_st_date)) # %i giống %d, in 1 số nguyên int

The examination will start from: 13/5/2020


**10, Nhập 1 số nguyên int và tính chúng với giá trị là n+nn+nnn. VD 5+55+555**!

**Write a Python program that accepts an integer (n) and computes the value of n+nn+nnn.**


![alt text](https://www.w3resource.com/w3r_images/python-basic-image-exercise-10.png)

In [0]:
a = int(input("Nhap 1 so nguyen: "))
n1 = int("%i"%a) # n=5, tạo 1 chuỗi là 5, rồi ép kiểu sang số int
n2 = int("%i%i"%(a,a)) # nn=55
n3 = int("%i%i%i"%(a,a,a)) # nnn=555
print(n1+n2+n3)


Nhap 1 so nguyen: 5
615


**11, In ra 1 tài liệu về Hàm xây dựng Python**

**Write a Python program to print the documents (syntax, description etc.) of Python built-in function(s).**



In [0]:
print(len.__doc__)
class Dog:
    """Your best friend."""

    def do_nothing(self):
        pass


print(Dog.__doc__)
# Your best friend. 

Return the number of items in a container.
Your best friend.


**12, In ra lịch gửi vào 1 tháng và năm**

**Write a Python program to print the calendar of a given month and year.**


Sử dụng module Calendar

![alt text](https://www.w3resource.com/w3r_images/python-basic-image-exercise-12.png)

In [0]:
import calendar
y = int(input("In put the year: "))
m = int(input("In put the month: "))
print(calendar.month(y, m))

In put the year: 9999
In put the month: 12
   December 9999
Mo Tu We Th Fr Sa Su
       1  2  3  4  5
 6  7  8  9 10 11 12
13 14 15 16 17 18 19
20 21 22 23 24 25 26
27 28 29 30 31



**13, Viết ra 1 chương trình 1 đoạn tài liệu**,

**Write a Python program to print the calendar of a given month and year.**



In [0]:
print("""a string that you "don't"" have to escape
This
is a ....... multi-line
heredoc string ---------> example""" )

a string that you "don't"" have to escape
This
is a ....... multi-line
heredoc string ---------> example


**14, Viết chương trình đếm số ngày giữa 2 ngày khác cách nhau**


**Write a Python program to calculate number of days between two dates.**


datetime.date(year, month, day) :

MINYEAR <= year <= MAXYEAR (MIN:1, MAX:9999)

1 <= month <= 12

1 <= day <= number of days in the given month and year

Ngoài vùng trả về lỗi ValueError

![alt text](https://www.w3resource.com/w3r_images/python-basic-image-exercise-14.png)

In [0]:
from datetime import date 
f_date = date(1996, 8, 7)
l_date = date(2020, 5, 13)
delta = l_date - f_date
print(delta.days)
print(delta)
print(f_date)

8680
8680 days, 0:00:00
1996-08-07


**15, Viết chương trình Python để lấy thể tích của một hình cầu có bán kính 6.**

**Write a Python program to get the the volume of a sphere with radius 6.**

![alt text](https://www.w3resource.com/w3r_images/python-volume-of-sphere-basic-exercise-15.png)

In [0]:
pi = 3.1415926535897931
r = 6.0
V = 4.0/3.0*pi*r**3
print('The volume of the shepre is: ', V)

The volume of the shepre is:  904.7786842338603


**16, Viết chương trình Python để lấy chênh lệch giữa một số đã cho và 17, nếu số lớn hơn 17 trả về gấp đôi số chênh lệch, nếu nhỏ hơn 17 lấy trị tuyệt đối**

**Write a Python program to get the difference between a given number and 17, if the number is greater than 17 return double the absolute difference.**
![alt text](https://www.w3resource.com/w3r_images/python-basic-image-exercise-16.png)

In [0]:
n = int(input("Nhap vao 1 so: "))
cl = n - 17
if n > 17:
  print(cl*2)
else:
  print(abs(cl))

"""def difference(n): C2: Viết hàm
  if n <= 17:
    return 17-n 
  else (n - 17) * 2
print(defference(22))
print(difference(14))"""

Nhap vao 1 so: 22
10


**17, Kiểm tra xem một số có nằm trong 100 của 1000 hoặc 2000 không**

**Write a Python program to test whether a number is within 100 of 1000 or 2000.**

![alt text](https://www.w3resource.com/w3r_images/python-abs-function-basic-exercise-17.png)

In [0]:
# Không hiểu đề bải lắm
def near_thousand(n):
  return ((abs(1000-n) <= 100) or (abs(2000 - n) <= 100)) #or: đúng khi 1 trong 2 đúng, trả về boolean kết quả true, false
print(near_thousand(1000))
print(near_thousand(900))
print(near_thousand(800))
print(near_thousand(2200))

True
True
False
False


**18, Viết chương trình Python để tính tổng của ba số đã cho, nếu các giá trị bằng nhau thì trả về ba lần tổng của chúng.**

**Write a Python program to calculate the sum of three given numbers, if the values are equal then return thrice of their sum.**



VD: 1 2 3 thì 1 ≠ 2 ≠ 3 thì 1 + 2 + 3 = 6

   3 3 3 thì 3 = 3 = 3 thì (3+3+3) * 3

![alt text](https://www.w3resource.com/w3r_images/python-basic-image-exercise-18.png)

In [0]:
def tinh_tong(x, y, z):
  sum = x + y + z
  if x != y != z:
    print(sum)
  else:
    print(sum*3)

tinh_tong(1,2,3)
tinh_tong(3,3,3)

""" def sum_thrice(x, y, z):      Cách 2
     sum = x + y + z
     if x == y == z:
      sum = sum * 3
     return sum
print(sum_thrice(1, 2, 3))
print(sum_thrice(3, 3, 3))"""

6
27


**19, Viết chương trình Python để lấy một chuỗi mới từ một chuỗi đã cho trong đó "Is" đã được thêm vào phía trước. Nếu chuỗi đã cho đã bắt đầu bằng "Is" thì trả về chuỗi không thay đổi.**

**Write a Python program to get a new string from a given string where "Is" has been added to the front. If the given string already begins with "Is" then return the string unchanged.**

![alt text](https://www.w3resource.com/w3r_images/python-basic-image-exercise-19.png)

In [0]:
def new_string(str):
  if len(str) >= 2 and str[:2] == "Is": # Điều kiện, đúng khi cả 2 đúng, độ dài phải lớn hơn 2 và index 0,1 là Is
    return str
  return "Is" + str # trả về hàm, nếu không return thì sẽ trả về None

print(new_string("Array"))
print(new_string("IsEmpty"))

IsArray
IsEmpty


**20, Viết chương trình Python để lấy một chuỗi là n (số nguyên không âm) sao chép của một chuỗi đã cho.**

**Write a Python program to get a string which is n (non-negative integer) copies of a given string.**



![alt text](https://www.w3resource.com/w3r_images/python-basic-image-exercise-20.png)


In [0]:
def larger_string(str, n):
  result = ""               # tạo 1 chuỗi resutlt rỗng
  for i in range(n):        # lặp số lần của n nhập vào tham số
    result = result + str   # gán kết quả vào result qua n lần lặp
  return result             # trả về hàm kết quả, nếu không return thì sẽ trả về None

print(larger_string('abc', 2))
print(larger_string('.py', 3))

abcabc
.py.py.py


**21, Tìm xem một số đã cho là chẵn hay lẻ, in ra một thông điệp phù hợp cho người dùng**

**Write a Python program to find whether a given number (accept from the user) is even or odd, print out an appropriate message to the user.**


![alt text](https://www.w3resource.com/w3r_images/check-even-number.png) ![alt text](https://www.w3resource.com/w3r_images/check-odd-number.png)

In [0]:
n = int(input("Nhap vao 1 so: "))
if n % 2 ==0: 
  print("So nay la so chan")
else:
  print("So nay la so le")

""" num = int(intput("Enter a number: ))
    mod = num % 2                           biến mod trả về số dư
    if mod > 0:                             số dư lớn hơn 0 thì là số lẻ
      print("This is an odd num number.")
    else:                                   số dư <= 0 thì là số chẵn
      print("This is an even number.)"""


Nhap vao 1 so: 10
So nay la so chan


**22, Đếm số 4 trong một danh sách nhất định**

**Write a Python program to count the number 4 in a given list.**


![alt text](https://www.w3resource.com/w3r_images/python-basic-image-exercise-22.png)

In [0]:
def list_count_4(nums):
  count = 0
  for num in nums:        # num duyệt từng phần tử trong list num, nếu num = 4 thì count được + thêm vào 1 
    if num == 4:
      count += 1
  return count
print(list_count_4([1, 4, 6, 7, 4]))
print(list_count_4([1, 4, 6, 4, 7, 4]))

2
3


**23, Viết chương trình Python để lấy n (số nguyên không âm), sao chép 2 ký tự đầu tiên của một chuỗi đã cho. Trả về n bản sao của toàn bộ chuỗi nếu độ dài nhỏ hơn 2.**

**Write a Python program to get the n (non-negative integer) copies of the first 2 characters of a given string. Return the n copies of the whole string if the length is less than 2.**

 ![alt text](https://www.w3resource.com/w3r_images/python-basic-image-exercise-23.png)

In [0]:
def substring_copy(str, n):
  flen = 2                      # biến flen kiểm tra độ dài, cố định là 2
  if flen > len(str):           # Nếu flen > độ dài chuỗi,
    flen = len(str)             # thì flen = độ dài chuỗi, độ dài là 1 thì flen = 1
  substr = str[:flen]           # Flen  < độ dài chuỗi, tạo biến substr, chuỗi con sẽ bằng index str[:2] hoặc str[:1]

  result = ""                   # tạo 1 biến result, khởi tạo ban đầu là rỗng, mục đích để gán substr vào result
  for i in range(n):            # duyệt và lặp n số lần
    result = result + substr    # gán substr vào result
  return result                 # kết quả trả về là result mới với số lần lặp đã hoàn thành
print(substring_copy('abcdef', 2))
print(substring_copy('p', 3));
print(substring_copy('', 3));



abab
ppp



**24, Kiểm tra 1 chữ cái có phải là nguyên âm hay không?**

**Write a Python program to test whether a passed letter is a vowel or not.**


![alt text](https://www.w3resource.com/w3r_images/python-basic-image-exercise-24.png)

In [0]:
def is_vowel(char):
  all_vowels = 'aeiou'        # biến all_vowels, chuỗi có chứa 5 nguyên âm
  return char in all_vowels   # trả về kiểm tra ký tự có nằm trong chuỗi all_vowels không? 
print(is_vowel('c'))
print(is_vowel('e'))

False
True


**25, Viết chương trình Python để kiểm tra xem một giá trị được chỉ định có được chứa trong một nhóm các giá trị hay không.**

**Write a Python program to check whether a specified value is contained in a group of values.**
![alt text](https://www.w3resource.com/w3r_images/python-basic-image-exercise-25.png)

In [0]:
def is_group_member(group_data, n):
  for value in group_data:              # duyệt từng phần tử trong group_data
    if n == value:                      # nếu n = value, tức n = 1 số trong những phần tử group_data
      return True                       # trả về True, tức là n có nằm trong group_data
  return False                          # trả về False, không nằm trong group_data

print(is_group_member([1, 5, 8, 3, 3], 3))
print(is_group_member([5, 8, 3], -1))

True
False


**26, Viết chương trình Python để tạo biểu đồ từ danh sách các số nguyên cho trước.**

**Write a Python program to create a histogram from a given list of integers.**


![alt text](https://www.w3resource.com/w3r_images/python-basic-image-exercise-26.png)

In [0]:
def histogram(items):
  for n in items:           # duyệt từng phần tử trong list items, n = giá trị từng phần tử của list
    output = ''             # tạo 1 biến là chuỗi output rỗng, dùng để chứa * vào chuỗi
    times = n               # tạo times, số lần thêm * vào output, times = n, n thì = giá trị trong 1 phần tử trong list
    while (times > 0):      # vòng lặp while, dừng khi điều kiện sai, số lần = 0 thì dừng vòng lặp
      output += '*'         # thêm * vào outputs, số lần thêm = times (n)
      times = times - 1     # thêm xong rồi thì - số lần đi 1 
    print(output)           # in ra ouput từng phần tử, không có return, trả về None
histogram([2,-1, 0, 3, 6, 5])


**


***
******
*****


**27, Viết chương trình Python để nối tất cả các phần tử trong danh sách thành một chuỗi và trả về nó.**

**Write a Python program to concatenate all elements in a list into a string and return it.**


![alt text](https://www.w3resource.com/w3r_images/python-basic-image-exercise-27.png)

In [0]:
def concatenate_list_data(list):
  result = ''                               # tạo 1 biến result rỗng, gán kết quả vào result
  for element in list:                      # duyệt từng phần tử trong list
    result += str(element)                  # thêm từng phần tử trong list vào result
  return result                             # trả về kết quả result
print(concatenate_list_data([1, 5, 12, 2]))


15122


**28, Viết chương trình Python để in ra tất cả các số chẵn từ danh sách các số đã cho theo cùng một thứ tự và dừng in nếu có bất kỳ số nào xuất hiện sau 237 trong chuỗi.**

**Write a Python program to print out all even numbers from a given numbers list in the same order and stop the printing if any numbers that come after 237 in the sequence.**

In [0]:
numbers = [    
    386, 462, 47, 418, 907, 344, 236, 375, 823, 566, 597, 978, 328, 615, 953, 345, 
    399, 162, 758, 219, 918, 237, 412, 566, 826, 248, 866, 950, 626, 949, 687, 217, 
    815, 67, 104, 58, 512, 24, 892, 894, 767, 553, 81, 379, 843, 831, 445, 742, 717, 
    958,743, 527
    ]
for x in numbers: # duyệt từng phần tử trong list
  if x == 237:    # điều kiên x = 237
    print(x)      # in ra x
    break;        # dừng, thoát khỏi vòng lặp và kết thúc chương trình
  elif x % 2 == 0:  # x chia hết cho 2, in ra x, x không chia hết cho 2, bỏ qua duyệt tiếp phần tử khác
    print(x)

386
462
418
344
236
566
978
328
162
758
918
237


**29, Viết chương trình Python để in ra một tập hợp chứa tất cả các màu từ color_list_1 không có trong color_list_2.**

**Write a Python program to print out a set containing all the colors from color_list_1 which are not present in color_list_2.**


![alt text](https://www.w3resource.com/w3r_images/python-basic-image-exercise-29.png)

In [0]:
color_list_1 = set(["White", "Black", "Red"]) # tạo 2 set
color_list_2 = set(["Red","Green"])
print(color_list_1.difference(color_list_2)) # hiệu 2 tập hợp, chỉ thực hiện được khi 2 biến là 2 tập hợp set

{'Black', 'White'}


**31, Viết chương trình Python sẽ chấp nhận đáy và chiều cao của một hình tam giác và tính diện tích.**

**Write a Python program that will accept the base and height of a triangle and compute the area.**


![alt text](https://www.w3resource.com/w3r_images/python-area-triangle-basic-exercise-30.png)

In [0]:
b = float(input("Input base: "))
h = float(input("input height: "))
area = 1/2 * b * h
print("Area = : ", area)

Input base: 20
input height: 40
Dien tich hinh tam giac la:  400.0


**32, Viết chương trình Python để lấy bội số chung nhỏ nhất (LCM) của hai số nguyên dương.**

**Write a Python program to compute the greatest common divisor (GCD) of two positive integers.**


![alt text](https://www.w3resource.com/w3r_images/python-basic-image-exercise-32.png)

In [0]:
def lcm(x, y):
  if x > y:
    z = x
  else: 
    z = y                                 # tạo 1 biến z được gán vào giá trị x hoặc y lớn nhất

  while(True):                            # tạo 1 vòng lặp, dừng đến khi gặp điều kiện break
    if((z % x == 0) and (z % y == 0)):    # điều kiện and, đúng khi cả 2 thỏa mãn, không thỏa mãn thì bỏ qua if
      lcm = z                             # gán z = lcm khi thỏa mãn điều kiện trên
      break                               # thoát khỏi vòng lặp
    z += 1                                # z + thêm 1 giá trị tăng dần, tiến tới khi nào z thỏa mãn điều kiện if ở trên
  return lcm                              # trả về giá trị lcm, lấy kết quả lcm được gán từ z
print(lcm(4, 6))
print(lcm(15, 17))


12
255


**33, Viết chương trình Python để tổng ba số nguyên đã cho. Tuy nhiên, nếu hai giá trị bằng nhau thì tổng sẽ bằng không.**

**Write a Python program to get the least common multiple (LCM) of two positive integers.**

![alt text](https://www.w3resource.com/w3r_images/python-basic-image-exercise-33.png)

In [0]:
def sum(x, y, z):
  if x == y or y == z or x == z: # Kiểm tra nếu x=y hoặc y=z hoặc x=z
    sum = 0                      # thì sum = 0
  else:                          # Nếu x, y, z khác nhau
    sum = x + y + z              # thì sum = x + y + z
  return sum                     # trả về kết quả giá trị hàm là sum
print(sum(2, 1, 2))
print(sum(3, 2, 2))
print(sum(2, 2, 2))
print(sum(1, 2, 3))

0
0
0
6


**34, Viết chương trình Python để tổng hai số nguyên đã cho. Tuy nhiên, nếu tổng từ 15 đến 20 thì nó sẽ trả về 20.**

**Write a Python program to sum of two given integers. However, if the sum is between 15 to 20 it will return 20.**


![alt text](https://www.w3resource.com/w3r_images/python-basic-image-exercise-34.png)

In [0]:
def sum(x, y):
    sum = x + y
    if sum in range(15, 20):
        return 20
    else:
        return sum

print(sum(10, 6))
print(sum(10, 2))
print(sum(10, 12))
print(sum(10, 10))

20
12
22
20


**35, Viết chương trình Python sẽ trả về true nếu hai giá trị nguyên đã cho bằng nhau hoặc tổng hoặc hiệu của chúng là 5.**

**Write a Python program which will return true if the two given integer values are equal or their sum or difference is 5.**


![alt text](https://www.w3resource.com/w3r_images/python-basic-image-exercise-35.png)

In [0]:
def sosanh(x, y):
  if x + y == 5:
    return True
  elif x - y == 5:
    return True
  elif x == y:
    return True
  else:
    return False

print(sosanh(7, 2))
print(sosanh(3, 2))
print(sosanh(2, 2))
print(sosanh(1, 2))
# cách 2:  if x == y or abs(x-y) == 5 or (x+y) == 5:   return True

True
True
True
False
